magento/magento2
# Drop indexer changelog tables on mview unsubscribe

### Description (*)
This pull request will remove the unused changelog tables when changing an index from 'update on schedule' to 'update on save'.

From what I can tell, this was the case before 6ae3c48d4df28a20590f7fe5d5deca472fd06084 was merged. I have tried to avoid any regressions here by marking the index as invalid when dropping the table, to catch any entries which might have otherwise been missed. It seems that 6ae3c48d4df28a20590f7fe5d5deca472fd06084 only protects missed changelog entries if the index is switched back to 'update on schedule'. The change introduced here solves that problem by always performing a full index (via normal minutely cron).

### Fixed Issues
- Fixes https://github.com/magento/ece-tools/issues/729
-  resolves magento/magento2#29789: Drop indexer changelog tables on mview unsubscribe

### Manual testing scenarios (*)
1. Set indexes to 'update on schedule'
1. Make a change which will trigger an entry in the changelog table (eg, update product name in admin)
1. Wait for normal minutely cron to run index
1. Observe data in `version_id` column in `mview_state` table.
1. Change indexes to 'update on save'
1. Observe data in `version_id` column in `mview_state` table.
1. Note that all changed indexes are now 'invalid' and will be fully indexed via normal minutely cron.

### Questions or comments
I am not confident that I have fully covered all changes here in unit tests. Please may I have some advice as to how I might need to adjust the unit tests to ensure no regressions.

### Contribution checklist (*)


### Resolved issues:
1. [x] resolves magento/magento2#29789: Drop indexer changelog tables on mview unsubscribe
